### DISCUSSION TITLE
Multiple threads within each MPI node

### DISCUSSION THREAD

**Discussion by deytonmoy000** [mpi4py/mpi4py#79] [Q&A] [answered]
Hi,
I would like to use multiple threads within each MPI node. 
Using "OMP_NUM_THREADS=nthreads", one can specify the number of threads (shared memory) to use within each MPI node but I would like to know how to control what the threads (rank) are performing similar to how one can control when a code is run with "mpirun -n" by using "COMM" class to obtain Size( getSize() ) and Rank ( getRank() ).

For eg. if I run "mpirun -np **ncores** OMP_NUM_THREADS= **nthreads** python3  test.py", is getSize() going to return "**ncores X nthreads**" or do I need to create a comm class variable from within the code running in the MPI node to obtain the Size( getSize() ) and Rank ( getRank() ) of the concerned node?

**Comment by dalcinl** [comment, marked as answer]
User-facing MPI (and thus mpi4py) knows nothing about threads. You have `MPI.Is_thread_main()`, but no other means of thread identification. `MPI.COMM_WORLD.Get_size()` will always return the number of PROCESSES (i.e. `-np` argument to `mpiexec`). Note that if you use OpenMP, the number of threads you specify are relevant to C extension modules built with OpenMP parallelism, but not the core Python interpreter itself or Python's `threading` module.

**Reply by deytonmoy000** [reply]
Thank you for the clarification. 
But is there any possibility/workaround to using multiple threads within each node with mpi4py?

**Reply by dalcinl** [reply]
Perhaps I was not clear enough. You can use multiple threads within each node. For example, if your code uses NumPy build with  a multithreaded BLAS/LAPACK implementation, some NumPy operations (e.g. dense linear solves) will run on multiple threads. If you use Python's `threading` module, you will be using multiple threads. The `mpi4py.futures` subpackage does use threading in its internal implementation. All was I was trying to explain is that whatever thread model you use, the way to handle these threads (e.g. how to identify one thread from the other, or getting the number of threads) is totally outside of the MPI or mpi4py real.

**Reply by deytonmoy000** [reply]
Thank you for the clarification.

**Reply by drmichaelt7777** [reply]
We usually set OMP_NUM_THREADS **before** we invoke the mpirun command to provide this info to OMP enabled apps. Such as

OMP_NUM_THREADS=4 mpirun -np ... other-mpi-options python -m _MPIenabled.py args_ ...

You can also permanently put this in the shell's environment ([ba]sh)

export OMP_NUM_THREADS=4 
mpirun ...
